test/e2e/apm/upgrade_test.go

```diff
@@ -0,0 +1,40 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+// +build apm e2e
+
+package apm
+
+import (
+	"testing"
+
+	"github.com/elastic/cloud-on-k8s/test/e2e/test"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/apmserver"
+	"github.com/elastic/cloud-on-k8s/test/e2e/test/elasticsearch"
+)
+
+func TestAPMServerVersionUpgradeToLatest8x(t *testing.T) {
+	srcVersion := test.Ctx().ElasticStackVersion
+	dstVersion := test.LatestSnapshotVersion8x
+
+	test.SkipInvalidUpgrade(t, srcVersion, dstVersion)
+
+	name := "apmserver-upgrade"
+	esBuilder := elasticsearch.NewBuilder(name).
+		WithVersion(srcVersion).
+		WithESMasterDataNodes(3, elasticsearch.DefaultResources)
+
+	apmServerBuilder := apmserver.NewBuilder(name).WithVersion(srcVersion).WithElasticsearchRef(esBuilder.Ref())
+
+	mutated := apmServerBuilder.WithVersion(dstVersion).WithElasticsearchRef(esBuilder.Ref()).WithMutatedFrom(&apmServerBuilder).WithoutIntegrationCheck()
+
+	test.RunMutations(
```

Comment:
This test does not succeed for me: 

```
{"log.level":"error","@timestamp":"2022-04-04T13:04:30.942Z","log.logger":"beater","log.origin":{"file.name":"beater/waitready.go","file.line":62},"message":"precondition 'apm integration installed' failed: e │
│ rror querying Elasticsearch for integration index templates: unexpected HTTP status: 404 Not Found ({\"error\":{\"root_cause\":[{\"type\":\"resource_not_found_exception\",\"reason\":\"index template matching  │
│ [logs-apm.error] not found\"}],\"type\":\"resource_not_found_exception\",\"reason\":\"index template matching [logs-apm.error] not found\"},\"status\":404}): to remediate, please install the apm integration:  │
│ https://ela.st/apm-integration-quickstart","service.name":"apm-server","ecs.version":"1.6.0"}     
```

Reply:
Interesting.  I'd been testing the 7.x -> 8.x upgrade path, not the 8.1 -> 8.2.  I'll verify that now as well.

Reply:
@pebrc weird, this isn't failing for me.  The error you're referencing above is supposed to be nullified by this:

https://github.com/elastic/cloud-on-k8s/pull/5470/files#diff-433f6bf9e8a193161ee44fc85c9a4325d740c1c1121a7989b24920b932bb6097R30

specifically the `.WithoutIntegrationCheck()` piece, which initially wasn't added to the original builder, but certainly should be, and I've added it now.  My test results are below:

```
--- PASS: TestVersionUpgradeSingleToLatest8x (331.51s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/K8S_should_be_accessible (0.08s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Elasticsearch_CRDs_should_exist (0.07s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Webhook_endpoint_should_not_be_empty (0.00s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Remove_Elasticsearch_if_it_already_exists (0.21s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Creating_an_Elasticsearch_cluster_should_succeed (0.08s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Elasticsearch_cluster_should_be_created (0.05s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_HTTP_certificate_authority_should_be_set_and_deployed (3.12s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_transport_certificate_authority_should_be_set_and_deployed (0.07s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/All_expected_Pods_should_eventually_be_ready (66.92s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_version_should_be_the_expected_one (0.11s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_services_should_be_created (0.18s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_services_should_have_endpoints (0.14s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Secrets_should_eventually_be_created (0.69s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_pods_should_eventually_have_a_certificate (0.17s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_cluster_health_should_eventually_be_green (6.16s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Elastic_password_should_be_available (0.05s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Elasticsearch_data_volumes_should_be_of_the_specified_type (0.06s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Cluster_should_be_annotated_with_its_UUID_once_bootstrapped (0.05s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_nodes_topology_should_eventually_be_the_expected_one (5.09s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_version_should_be_the_expected_one#01 (2.55s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_endpoint_should_eventually_be_reachable (2.60s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Verify_TLS_CA_cert_on_transport_layer_is_the_expected_one (2.24s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Add_some_data_to_the_cluster_before_starting_the_mutation (4.91s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Retrieve_Elasticsearch_cluster_UUID_for_comparison_purpose (2.39s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Retrieve_Objects_metadata.Generation,_and_status.ObservedGeneration_for_comparison_purpose (0.11s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Annotate_Pods_with_a_hash_of_their_Builder_spec (0.21s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Applying_the_Elasticsearch_mutation_should_succeed (3.23s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_HTTP_certificate_authority_should_be_set_and_deployed#01 (0.05s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_transport_certificate_authority_should_be_set_and_deployed#01 (0.05s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/All_expected_Pods_should_eventually_be_ready#01 (149.48s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_version_should_be_the_expected_one#02 (0.12s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_services_should_be_created#01 (0.15s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_services_should_have_endpoints#01 (0.15s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Secrets_should_eventually_be_created#01 (0.74s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_pods_should_eventually_have_a_certificate#01 (0.15s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_cluster_health_should_eventually_be_green#01 (6.21s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Elastic_password_should_be_available#01 (0.08s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Elasticsearch_data_volumes_should_be_of_the_specified_type#01 (0.07s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Cluster_should_be_annotated_with_its_UUID_once_bootstrapped#01 (0.06s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_nodes_topology_should_eventually_be_the_expected_one#01 (2.91s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_version_should_be_the_expected_one#03 (2.49s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/ES_endpoint_should_eventually_be_reachable#01 (2.60s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Verify_TLS_CA_cert_on_transport_layer_is_the_expected_one#01 (2.29s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Cluster_UUID_should_have_been_preserved (2.39s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/metadata.generation_and_status.observedGeneration_may_have_been_incremented_from_previous_state,_and_should_be_equal (0.12s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Data_added_initially_should_still_be_present (2.46s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Deleting_Elasticsearch_should_return_no_error (0.07s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Elasticsearch_should_not_be_there_anymore (0.05s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Elasticsearch_pods_should_eventually_be_removed (55.20s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/PVCs_should_eventually_be_removed (0.08s)
    --- PASS: TestVersionUpgradeSingleToLatest8x/Soft-owned_secrets_should_eventually_be_removed (0.17s)
 ```
 
 How were you running the test?  Mine was something like
 
 ```
 make e2e-local TESTS_MATCH=TestVersionUpgradeSingleToLatest8x E2E_STACK_VERSION=8.1.0
 ```

Reply:
The test I am running is a different one: `TestAPMServerVersionUpgradeToLatest8x`

Reply:
@pebrc replicating this is causing me nothing but problems.  Initially the 8.2.0-SNAPSHOT sha discrepancy between versions issue hit me, and now testing on another cluster, I'm getting failures here
```
=== RUN   TestAPMServerVersionUpgradeToLatest8x/Events_should_eventually_show_up_in_Elasticsearch#01
```
will no errors in apmserver, and nothing clear in elasticsearch.  I'm still trying to get to the bottom of it.

Reply:
After some debugging this morning, I've found the cause of this failure in the e2e tests.  I'm working through the cleanest way to solve for it.

Reply:
I've solved for this issue by tracking the count of documents in both the metric, and error index used by APM Server prior to ingested an event, and checking the new count relative to the previous count.